clang/lib/CIR/Dialect/Transforms/LoweringPrepareAArch64CXXABI.cpp

```diff
@@ -0,0 +1,176 @@
+
+  //=======================================
+  // Argument was on the stack
+  //=======================================
+  builder.setInsertionPointToEnd(onStackBlock);
+  auto stackP = builder.createGetMemberOp(loc, valist, "stack", 0);
+  auto stack = builder.create<mlir::cir::LoadOp>(loc, stackP);
+  auto ptrDiffTy =
+      mlir::cir::IntType::get(builder.getContext(), 64, /*signed=*/false);
+  auto eight = builder.create<mlir::cir::ConstantOp>(
+      loc, ptrDiffTy, mlir::cir::IntAttr::get(ptrDiffTy, 8));
+  auto i8Ty = IntegerType::get(builder.getContext(), 8);
+  auto i8PtrTy = PointerType::get(builder.getContext(), i8Ty);
+  auto castStack = builder.createBitcast(stack, i8PtrTy);
+  // Write the new value of __stack for the next call to va_arg
+  auto newStackAsi8Ptr = builder.create<mlir::cir::PtrStrideOp>(
+      loc, castStack.getType(), castStack, eight);
+  auto newStack = builder.createBitcast(newStackAsi8Ptr, stack.getType());
+  builder.createStore(loc, newStack, stackP);
+  builder.create<mlir::cir::BrOp>(loc, mlir::ValueRange{stack}, newEndBlock);
+
+  //=======================================
+  // Argument was in registers
+  //=======================================
+  // Now we emit the code for if the argument was originally passed in
+  // registers. First start the appropriate block:
+  builder.setInsertionPointToEnd(inRegBlock);
+  auto regTopP = builder.createGetMemberOp(loc, valist,
+                                           isFloatingType ? "vr_top" : "gr_top",
+                                           isFloatingType ? 2 : 1);
+  auto regTop = builder.create<mlir::cir::LoadOp>(loc, regTopP);
+  auto castRegTop = builder.createBitcast(regTop, i8PtrTy);
+  auto resAsInt8P = builder.create<mlir::cir::PtrStrideOp>(
+      loc, castRegTop.getType(), castRegTop, offs);
+  auto resAsVoidP = builder.createBitcast(resAsInt8P, regTop.getType());
+  builder.create<mlir::cir::BrOp>(loc, mlir::ValueRange{resAsVoidP},
+                                  newEndBlock);
+
+  // generate additional instructions for end block
+  builder.setInsertionPoint(op);
+  newEndBlock->addArgument(stack.getType(), loc);
+  auto resP = newEndBlock->getArgument(0);
+  assert(resP.getType().isa<mlir::cir::PointerType>());
+  auto opResPTy = PointerType::get(builder.getContext(), opResTy);
+  auto castResP = builder.createBitcast(resP, opResPTy);
+  auto res = builder.create<mlir::cir::LoadOp>(loc, castResP);
+  return res.getResult();
+}
+
+mlir::Value LoweringPrepareAArch64CXXABI::lowerVAArg(CIRBaseBuilderTy &builder,
+                                                     mlir::cir::VAArgOp op) {
+
+  if (Kind == clang::TargetCXXABI::GenericAArch64) {
+    return lowerGenericAArch64VAArg(builder, op);
+  }
+  // Return empty value here so CIR Lowering don't do anything.
+  // This would leave the responsibility to LLVM Lowering to handle cir.va_arg
+  // which could be better if algorithm is simple.
+  return mlir::Value();
```

Comment:
Unfortunately there won't be a case where a generic version would be target independent, please make this into a `llvm_unreachable("NYI");`

Reply:
same argument here. We won't cover up anything when we do nothing here. LowerToLLVM will throw error when it sees cir.var_arg

Reply:
I don't us to rely on other paths to say something about a specific implementation, it's possible they change, and all of a sudden we were relying on something and boom. Both places should have the NYI assertions.